src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,145 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (first == null) {
+            first = new Node<>(null, value, null);
+            last = first;
+        } else {
+            Node<T> newNode = new Node<>(last, value, null);
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index has not been found");
+        } else if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, first);
+            first.prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> nodeByIndex = getNodesIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodesIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> actualNode = getNodesIndex(index);
+        T oldValue = actualNode.item;
+        actualNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlink(getNodesIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> actualNode;
+        for (actualNode = first; actualNode != null; actualNode = actualNode.next) {
+            if (actualNode.item == object || object != null && object.equals(actualNode.item)) {
+                unlink(actualNode);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isIndexValid(int index) {
+        return index <= size && index >= 0;
+    }
+
+    private Node<T> getNodesIndex(int index) {
+        if (index == size || !isIndexValid(index)) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> actualNode;
+        int i;
+        if (index < size / 2) {
+            actualNode = first;
+            for (i = 0; i < index; i++) {
+                actualNode = actualNode.next;
+            }
+            return actualNode;
+        } else {
```

Comment:
we can remove `else` and write all code after `if` block
```suggestion
        } 
```

Reply:
Fix